keps/sig-architecture/20190412-conformance-behaviors.md

```diff
@@ -0,0 +1,342 @@
+feasible to automate the creation of simple tests for these behaviors, along
+with the behavior descriptions in the `spec.yaml`. In some cases a complete test
+may not be easy to generate, but a skeleton may be created that can be be
+converted into a valid test with minimal effort.
+
+For these tests, the input is a set of manifests that are applied to the
+cluster, along with a set of conditions that are expected to be realized within
+a specified timeframe. The test framework will apply the manifests, and monitor
+the cluster for the conditions to occur; if they do not occur within the
+timeframe, the test will fail.
+
+For each Spec object, scaffolding can be defined to include the following tests:
+
+* Creation and read of the resource with only required fields specified.
+  * API functions as expected: Resource is created and may be read, and defaults
+    are set. This is mechanical and can be completely generated.
+  * Cluster behaves as expected. This cannot be generated, but a skeleton can be
+    generated that allows test authors to evaluate the condition of the cluster
+    to make sure it meets the expectations.
+* Deletion of the resource. This may be mostly mechanical but if there are side-
+  effects, such as garbage collection of related resources, we may want to have
+  manually written evaluation here as well.
+* Creation of resource with each field set, and update of each mutable field.
+  * For each mutable field, apply a patch to the based resource definition
+    before creation (for create tests), or after creation (for update tests).
+  * Evaluate that the API functions as expected; this is mechanical and
+    generated.
+  * Evaluate that the cluster behaves as expected. In some cases this may be
+    able to re-use the same evaluation function used during the creation tests,
+    but often it will require hand-crafted code to test the conditions.
+    Nonetheless, the scaffolding can be generated, minimizing the effort needed
+    to implement the test.
+
+Additional, handwritten tests will be needed that modify the resource in
+multiple ways and evaulate the behavior. The scaffolding must be built such that
+the same process is used for these tests. The test author must only need to
+define:
+* A patch to apply to create or update the resource.
+* A function to evaluate the effect of the API call.
+
+With those two, the same creation and update scaffolding defined for individual
+field updates can be reused.
+
+### Coverage Tooling
+
+In order to tie behaviors back to the tests that are generated, including
+existing e2e tests that already cover behaviors, new tags with behavior IDs will
+be added to conformance tests. Using the existing conformance framework
+mechanism allows incremental adoption of this proposal. Thus, rather than a new
+conformance framework function, test authors will indicate the behaviors covered
+by their tests with a tag in the `framework.ConformanceIt` call.
```

Comment:
ideas:
- `Testname: behavior id`
- `ConformanceIt("behavior id", "should function for node-pod communication:", func...`
- Change Conformance tag to be like our Feature tag, `[Conformance:behavior id]` (but this would break back compat)